File: src/main/java/com/epicsquid/ranchersdelight/machines/mayo/MayoMachineBlock.java
package com.epicsquid.ranchersdelight.machines.mayo;

import com.epicsquid.ranchersdelight.init.RanchersDelightBlockEntities;
import com.epicsquid.ranchersdelight.machines.base.TickableEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.EntityBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityTicker;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraftforge.network.NetworkHooks;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public class MayoMachineBlock extends Block implements EntityBlock {

	public MayoMachineBlock(Properties pProperties) {
		super(pProperties);
	}

	@Override
	protected void createBlockStateDefinition(@NotNull StateDefinition.Builder<Block, BlockState> pBuilder) {
		super.createBlockStateDefinition(pBuilder);
		pBuilder.add(BlockStateProperties.POWERED);
		pBuilder.add(BlockStateProperties.FACING);
	}

	@Nullable
	@Override
	public BlockState getStateForPlacement(@NotNull BlockPlaceContext pContext) {
		return super.getStateForPlacement(pContext)
				.setValue(BlockStateProperties.POWERED, false)
				.setValue(BlockStateProperties.FACING, Direction.UP);
	}

	@Nullable
	@Override
	public BlockEntity newBlockEntity(@NotNull BlockPos pos, @NotNull BlockState state) {
		return RanchersDelightBlockEntities.MAYONNAISE_MACHINE.create(pos, state);
	}

	@Nullable
	@Override
	public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level pLevel, BlockState pState, BlockEntityType<T> pBlockEntityType) {
		return TickableEntity::tick;
	}

	@Override
	public InteractionResult use(BlockState pState, Level pLevel, BlockPos pPos, Player pPlayer, InteractionHand pHand, BlockHitResult pHit) {
		if (pLevel.isClientSide) {
			return InteractionResult.SUCCESS;
		}
		BlockEntity be = pLevel.getBlockEntity(pPos);
		if (be instanceof MayoMachineBlockEntity mayoBe) {
			NetworkHooks.openScreen((ServerPlayer) pPlayer, mayoBe, mayoBe::sendToMenu);
		}
		return InteractionResult.SUCCESS;
	}
}

File: src/main/kotlin/dev/epicsquid/ranchersdelight/machines/mayo/MayoMachineBlock.kt
package dev.epicsquid.ranchersdelight.machines.mayo

import dev.epicsquid.ranchersdelight.machines.base.TickableEntity
import dev.epicsquid.ranchersdelight.registry.BlockEntityRegistry
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction.UP
import net.minecraft.world.item.context.BlockPlaceContext
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.EntityBlock
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityTicker
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.StateDefinition.Builder
import net.minecraft.world.level.block.state.properties.BlockStateProperties

class MayoMachineBlock(
	props: Properties
) : Block(props), EntityBlock {
	override fun createBlockStateDefinition(builder: Builder<Block, BlockState>) {
		super.createBlockStateDefinition(builder)
		builder.add(BlockStateProperties.POWERED)
		builder.add(BlockStateProperties.FACING)
	}

	override fun getStateForPlacement(context: BlockPlaceContext): BlockState? =
		super.getStateForPlacement(context)
			?.setValue(BlockStateProperties.POWERED, false)
			?.setValue(BlockStateProperties.FACING, UP)

	override fun newBlockEntity(pos: BlockPos, state: BlockState): BlockEntity? {
		return BlockEntityRegistry.mayonnaiseMachine.create(pos, state)
	}

	override fun <T : BlockEntity?> getTicker(
		tickerLevel: Level,
		tickerState: BlockState,
		tickerEntityType: BlockEntityType<T>
	): BlockEntityTicker<T> = BlockEntityTicker(TickableEntity::tick)
}
